src/natcap/invest/sdr/sdr_core.pyx

```diff
@@ -471,62 +513,76 @@ def calculate_sediment_deposition(
                         neighbor_flow_weight = (
                             neighbor_flow_val >> (inflow_offsets[j]*4)) & 0xF
                         if neighbor_flow_weight > 0:
                             f_j = f_raster.get(neighbor_col, neighbor_row)
+                            if f_j == sediment_deposition_nodata:
+                                continue
+                            # sum up the neighbor's flow dir values in each direction
+                            # flow dir values are relative to the total
                             neighbor_flow_sum = 0
                             for k in range(8):
                                 neighbor_flow_sum += (
                                     neighbor_flow_val >> (k*4)) & 0xF
+                            # get the proportion of the neighbor's flow that
+                            # flows into the original pixel
                             p_val = neighbor_flow_weight / neighbor_flow_sum
+                            # add the neighbor's flux value, weighted by the flow proportion
                             f_j_weighted_sum += p_val * f_j
 
-                    # calculate the differential downstream change in sdr
-                    # from this pixel
-                    downstream_sdr_weighted_sum = 0.0
+                    # calculate sum of SDR values of immediate downstream neighbors,
+                    # weighted by proportion of flow into each neighbor
+                    # (sum over k ∈ K of SDR_k * p(i,k) in the equation above)
+                    downstream_sdr_weighted_sum = 0
                     flow_val = <int>mfd_flow_direction_raster.get(
                         global_col, global_row)
-                    flow_sum = 0.0
+                    flow_sum = 0
                     for k in range(8):
                         flow_sum += (flow_val >> (k*4)) & 0xF
 
+                    # iterate over the neighbors again
                     for j in range(8):
+                        # skip if neighbor is outside the raster boundaries
                         neighbor_row = global_row + ROW_OFFSETS[j]
                         if neighbor_row < 0 or neighbor_row >= n_rows:
                             continue
                         neighbor_col = global_col + COL_OFFSETS[j]
                         if neighbor_col < 0 or neighbor_col >= n_cols:
                             continue
-                        # if this direction flows out, add to weighted sum
+                        # if it is a downstream neighbor, add to the sum and
+                        # check if it can be pushed onto the stack yet
                         flow_weight = (flow_val >> (j*4)) & 0xF
                         if flow_weight > 0:
                             sdr_j = sdr_raster.get(neighbor_col, neighbor_row)
-                            if sdr_j == 0.0:
+                            if sdr_j == sdr_nodata:
+                                continue
```

Comment:
This is the only place where `continue`ing on nodata could have an impact beyond 

The first part here is summing up the `downstream_sdr_weighted_sum`. The effect is the same because before, `if sdr_j == sdr_nodata` we would add `0`. Now, we `continue` and don't add anything.

The second part is checking if the neighbor `j` can be pushed onto the stack yet. If`sdr_j == sdr_nodata`, then the pixel `j` will never get pushed onto the stack. That's correct because we can't calculate the outputs on pixel `j` if `sdr_j` is undefined. This may be redundant with the [check at the end](https://github.com/emlys/invest/blob/1b71cd469ba68e10f103bdf38e416113cf50856c/src/natcap/invest/sdr/sdr_core.pyx#L616) but it does make the intent clearer.

The potential impact of this is that, if you have a pixel whose upstream neighbors all have undefined SDR, I think it would never get calculated. I believe that's okay but not sure.

Reply:
> This may be redundant with the check at the end but it does make the intent clearer.

It'll also break out of unnecessary computation earlier, which is a very good thing in these low-level functions

>The potential impact of this is that, if you have a pixel whose upstream neighbors all have undefined SDR, I think it would never get calculated. I believe that's okay but not sure.

Wouldn't we have a contradiction if this were true?  It seems to me that the only way we'd have undefined SDR for upstream pixels _U_ of pixel _p_ would be for upstream pixels _U_ to be in a distinct hydrological region ... but if they're in a distinct hydrological region (because they flow to a nonexistent stream and not the same stream _p_ flows to) they'd have to be downstream of _p_, right?  

I _think_ we might be OK here, but one thing I might suggest is to try to break it!  Can you construct a case (one that the model would produce given reasonable inputs) that exercises what you're concerned about?  If you can, that would be a very good start to a test case to make sure we handle it correctly!

Reply:
If we assume that this function will only be called on the rasters produced by `sdr.py`, I think you're right. In the model, the SDR raster should have data everywhere that the flow_dir raster has data.

If `sdr_core.calculate_sediment_deposition` were called with an SDR input that _didn't_ come from the model, you can construct a scenario where this happens:
```
pixel names     flow dir        SDR        E'
 a  d  g        -> -> ->     .5 .5 .5    1  1  1
 b  e  h        -> -> ->     .5 -1 .5    1  1  1
 c  f  i        -> -> ->     .5 -1 .5    1  1  1
```
Pixels a, d, and g are seed pixels. Pixels will be pushed onto the stack in the order:  `a, d, g, b, c`. 
`e` and `f` aren't ever pushed to the stack because SDR is nodata there. And so `h` and `i` are never computed either.

I think this would be correct:
- `dR` is defined in terms of SDR
- therefore `dR` should have nodata where SDR has nodata
- `f` is defined in terms of `dR`
- therefore, `f` should have nodata where SDR has nodata
- `f` is also defined recursively in terms of the immediate upstream neighbors
- therefore, if a pixel `i`'s has an upstream neighbor `j`, and `f_j` is nodata, then `f_i` should be nodata
- therefore, if a pixel `i`'s has an upstream neighbor `j`, and `SDR_j` is nodata, then `f_i` should be nodata
- therefore, `f` should be nodata on all pixels downstream of a nodata SDR pixel

You could argue that pixels `b` and `c` should also be undefined in the output, because `r` and `f` are defined in terms of SDR for all _downstream_ neighbors of `i`, which is `nodata` for these pixels.

Reply:
I made that change in c99bc44 just so you can see what I mean. I don't know if that makes sense to do.

Reply:
>If `sdr_core.calculate_sediment_deposition` were called with an SDR input that didn't come from the model, you can construct a scenario where this happens

Thanks for providing this level of detail around a scenario that would exercise this.  And even though it is possible (as you demonstrate) that we could end up with some undefined behavior, I'd like to argue for the simpler solution of assuming that all of this function's inputs will come from the SDR model, clarifying this by renaming `calculate_sediment_deposition` to `_calculate_sediment_deposition` (and doing the same with the other `sdr_core` cython function) because:

1. All other functions used in the SDR task graph are likewise private
2. Private functions in python imply by convention that they are for internal use only and should not be used directly outside of the context of the model's `execute`
3. All of this would help to enforce that the contract we're exposing to the user is at the `execute` level and not for each individual component.

I suppose it's always possible for someone to call a function in a way they aren't supposed to and with data that aren't formatted correctly, but I'm just really hesitant to try to support something that shouldn't happen under normal model operation.

Reply:
@emlys any thoughts on the above?

Reply:
Whoops, didn't see this! This makes sense to me. I reset c99bc44 after committing it so those changes aren't in here now. I did just add to the docstring explaining this assumption.